crates/sui-source-validation/src/lib.rs

```diff
@@ -617,15 +617,37 @@ fn download_and_compile(
         // Check the platform and proceed if we can download a binary. If not, the user should follow error instructions to sideload the binary.
         let platform = detect_platform(&root, compiler_version, &dest_canonical_binary)?;
         // Download if binary does not exist.
-        let url = format!("https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz", compiler_version, compiler_version, platform);
+        let mainnet_url = format!(
+            "https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz",
+            compiler_version, compiler_version, platform
+        );
 
         println!(
-            "{} compiler @ {} (this may take a while)",
+            "{} mainnet compiler @ {} (this may take a while)",
             "DOWNLOADING".bold().green(),
             compiler_version.yellow()
         );
 
-        let mut response = ureq::get(&url).call()?.into_reader();
+        let mut response = match ureq::get(&mainnet_url).call() {
+	    Ok(response) => response,
+	    Err(ureq::Error::Status(404, _)) => {
+		    println!(
+			"{} sui mainnet compiler {} not available, attempting to download testnet compiler release...",
+			"WARNING".bold().yellow(),
+			compiler_version.yellow()
+
+		    );
+		    println!(
+			"{} testnet compiler @ {} (this may take a while)",
+			"DOWNLOADING".bold().green(),
+			compiler_version.yellow()
+		    );
+		    let testnet_url = format!("https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz", compiler_version, compiler_version, platform);
+		    ureq::get(&testnet_url).call()?
```

Comment:
Wouldn't this return a not so nice error in case of failure which does not align visually with the previous print statements and warnings??

Reply:
To make sure I follow, you mean the uglier red error message like this:

![Screenshot 2024-01-26 at 5 47 32 PM](https://github.com/MystenLabs/sui/assets/888624/5a04e055-f35e-4529-926f-4d1884950eb9)

Instead of something prettier like "ERROR: blah blah"?

If yes, then I think we can make _some_ of these errors prettier (like the binary download one), but the aggregated source verification errors have kind of always been that way. Those could be improved too, but so far we haven't focused much more on the visual presentation here.

Will merge this, but please follow up with comment and I'll address if I follow the ask here :-)